skiko/src/awtMain/kotlin/org/jetbrains/skiko/Actuals.awt.kt

```diff
@@ -41,7 +41,17 @@ internal actual fun makeDefaultRenderFactory(): RenderFactory =
     }
 
 internal actual fun URIHandler_openUri(uri: String) {
-    Desktop.getDesktop().browse(URI(uri))
+    val desktop = Desktop.getDesktop()
+    if (desktop.isSupported(Desktop.Action.BROWSE)) {
+        desktop.browse(URI(uri))
+    } else when (hostOs) {
+        OS.Linux -> {
+            Runtime.getRuntime().exec(arrayOf("xdg-open", URI(uri).toString()))
```

Comment:
It should parse as a URL, not just a URI.

Reply:
Done. To preserve the existing exception behavior, I first parse as URI and then as URL. This way the user code only gets URI exceptions as in all other cases, and never the different `MalformedURLException`.

Reply:
It shouldn't throw `UnsupportedOperationException` in this case, because the operation is supported; it's the argument that's wrong. It should throw `IllegalArgumentException` (per `Desktop.browse` docs).

Reply:
> It should throw IllegalArgumentException (per Desktop.browse docs).

I didn't realize the URI has to be a valid URL, based on the method name `URIHandler_openUri` I would think it can handle URIs and not only specifically URLs. Why is not called `URLHandler_openUrl` if other valid URIs are considered illegal arguments? Confused me. Updated.

Reply:
Not sure if we can delete the `URI(uri)` validation if we are validating the URL anyway. I would think we can delete it, but maybe there is some edge case scenario where a valid URL is somehow not a valid URI? Not sure.

Reply:
`URIHandler_openUri` is supposed to handle URIs too, but `xdg-open` can't, and I prefer to sanitize the input as much as possible before passing it to a 3rd-party utility.

> Not sure if we can delete the URI(uri) validation I added on line 51 if we are validating the URL anyway. I would think we can delete it, but maybe there is some edge case scenario where a valid URL is somehow not a valid URI? Not sure.

We probably can. I understand URLs are a subset of URIs.

Reply:
> URIHandler_openUri is supposed to handle URIs too, but xdg-open can't, and I prefer to sanitize the input as much as possible before passing it to a 3rd-party utility.

If `URIHandler_openUri` is supposed to handle URIs, and the user has passed in a valid URI, then is it really an `IllegalArgumentException` that `xdg-open` doesn't support that URI (because it's not a URL)? This was my thinking here.

Reply:
I see what you mean. Maybe throw `IOException` then (or pass through the `MalformedUrlException`; it's a subclass of `IOException`, strangely).

Reply:
All right, `MalformedUrlException` now goes to the caller (in the rare case a valid URI is not a valid URL).